File: src/commands/import.ts
import {
  type ChatInputCommandInteraction,
  type Invite,
  MessageFlags,
  SlashCommandBuilder,
} from "discord.js";
import { count, eq } from "drizzle-orm";
import type { Command } from "../client.js";
import { db } from "../db/index.js";
import { guilds, inviteRoles, trackedInvites } from "../db/schema.js";
import { logEmbed, sendLogMessage, writeAuditLog } from "../lib/auditLog.js";
import { brandEmbed, errorEmbed, successEmbed } from "../lib/embeds.js";
import { updateCachedInvite } from "../lib/inviteCache.js";
import { requireGuild } from "../lib/permissions.js";
import { logger } from "../logger.js";

interface CsvRow {
  inviteCode: string;
  primarySource: string;
  secondarySource: string;
  creatorId: string;
  createdAt: Date;
  roleIds: string[];
}

function parseCsv(content: string): CsvRow[] {
  const lines = content.trim().split("\n");
  if (lines.length < 2) return [];

  // Skip header line
  return lines.slice(1).map((line) => {
    const parts = line.split(",").map((s) => s.trim());
    const roleIds = parts[5]
      ? parts[5]
          .split(";")
          .map((r) => r.trim())
          .filter(Boolean)
      : [];

    return {
      inviteCode: parts[0],
      primarySource: parts[1],
      secondarySource: parts[2],
      creatorId: parts[3],
      createdAt: parts[4] ? new Date(parts[4]) : new Date(),
      roleIds,
    };
  });
}

const command: Command = {
  data: new SlashCommandBuilder()
    .setName("import")
    .setDescription("Import existing invite links into tracking")
    .setDefaultMemberPermissions(0x8)
    .addSubcommand((sub) =>
      sub
        .setName("single")
        .setDescription("Import a single existing invite")
        .addStringOption((opt) =>
          opt
            .setName("code")
            .setDescription("The invite code")
            .setRequired(true),
        )
        .addStringOption((opt) =>
          opt
            .setName("primary_source")
            .setDescription("Primary source")
            .setRequired(true),
        )
        .addStringOption((opt) =>
          opt
            .setName("secondary_source")
            .setDescription("Secondary source")
            .setRequired(true),
        )
        .addRoleOption((opt) =>
          opt.setName("role").setDescription("Role to auto-assign on join"),
        ),
    )
    .addSubcommand((sub) =>
      sub
        .setName("csv")
        .setDescription("Bulk import from a CSV file")
        .addAttachmentOption((opt) =>
          opt
            .setName("file")
            .setDescription("CSV file to import")
            .setRequired(true),
        ),
    ),

  async execute(interaction: ChatInputCommandInteraction) {
    const guildId = await requireGuild(interaction);
    if (!guildId) return;
    const subcommand = interaction.options.getSubcommand();

    // Ensure guild exists
    await db.insert(guilds).values({ id: guildId }).onConflictDoNothing();

    if (subcommand === "single") {
      await handleSingleImport(interaction, guildId);
    } else if (subcommand === "csv") {
      await handleCsvImport(interaction, guildId);
    }
  },
};

async function handleSingleImport(
  interaction: ChatInputCommandInteraction,
  guildId: string,
) {
  const code = interaction.options
    .getString("code", true)
    .replace("https://discord.gg/", "")
    .replace("discord.gg/", "");
  const primarySource = interaction.options.getString("primary_source", true);
  const secondarySource = interaction.options.getString(
    "secondary_source",
    true,
  );
  const roleOption = interaction.options.getRole("role");

  const guild = interaction.guild;
  if (!guild) return;

  // Verify invite exists in guild
  let discordInvite: Invite | undefined;
  try {
    discordInvite = await guild.invites.fetch(code);
  } catch {
    await interaction.reply({
      embeds: [errorEmbed(`Invite \`${code}\` not found in this server.`)],
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  if (!discordInvite) {
    await interaction.reply({
      embeds: [errorEmbed(`Invite \`${code}\` not found in this server.`)],
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  // Check if already tracked
  const existing = await db
    .select()
    .from(trackedInvites)
    .where(eq(trackedInvites.inviteCode, code))
    .limit(1);

  if (existing.length > 0) {
    await interaction.reply({
      embeds: [errorEmbed(`Invite \`${code}\` is already being tracked.`)],
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  const uses = discordInvite.uses ?? 0;

  // Insert
  const [inserted] = await db
    .insert(trackedInvites)
    .values({
      guildId,
      inviteCode: code,
      channelId: discordInvite.channel?.id ?? "",
      primarySource,
      secondarySource,
      createdBy: interaction.user.id,
      uses,
    })
    .returning();

  if (roleOption && inserted) {
    await db.insert(inviteRoles).values({
      trackedInviteId: inserted.id,
      roleId: roleOption.id,
    });
  }

  await updateCachedInvite(guildId, code, uses);

  await writeAuditLog(guildId, "invite_imported", interaction.user.id, {
    inviteCode: code,
    primarySource,
    secondarySource,
  });

  const embed = successEmbed()
    .setTitle("Invite Imported!")
    .setDescription(
      `Imported \`${code}\` with **${uses}** existing uses.\n**${primarySource}** \u2192 **${secondarySource}**`,
    );

  await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
}

async function handleCsvImport(
  interaction: ChatInputCommandInteraction,
  guildId: string,
) {
  const guild = interaction.guild;
  if (!guild) return;

  const attachment = interaction.options.getAttachment("file", true);

  if (!attachment.name.endsWith(".csv")) {
    await interaction.reply({
      embeds: [errorEmbed("Please upload a `.csv` file.")],
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  await interaction.deferReply({ flags: MessageFlags.Ephemeral });

  // Download CSV content
  const response = await fetch(attachment.url);
  const content = await response.text();

  const rows = parseCsv(content);
  if (rows.length === 0) {
    await interaction.editReply({
      embeds: [errorEmbed("CSV file is empty or has invalid format.")],
    });
    return;
  }

  // Check limit
  const [guildConfig] = await db
    .select()
    .from(guilds)
    .where(eq(guilds.id, guildId))
    .limit(1);

  const [{ value: currentCount }] = await db
    .select({ value: count() })
    .from(trackedInvites)
    .where(eq(trackedInvites.guildId, guildId));

  const maxLinks = guildConfig?.maxLinks ?? 130;

  let imported = 0;
  let skipped = 0;
  let errors = 0;
  const skippedReasons: string[] = [];

  for (const row of rows) {
    if (currentCount + imported >= maxLinks) {
      skipped += rows.length - imported - skipped - errors;
      skippedReasons.push("Link limit reached");
      break;
    }

    try {
      // Check if already tracked
      const existing = await db
        .select()
        .from(trackedInvites)
        .where(eq(trackedInvites.inviteCode, row.inviteCode))
        .limit(1);

      if (existing.length > 0) {
        skipped++;
        continue;
      }

      // Verify invite exists
      let discordInvite: Invite | undefined;
      try {
        discordInvite = await guild.invites.fetch(row.inviteCode);
      } catch {
        skipped++;
        skippedReasons.push(`\`${row.inviteCode}\` — not found`);
        continue;
      }

      if (!discordInvite) {
        skipped++;
        skippedReasons.push(`\`${row.inviteCode}\` — not found`);
        continue;
      }

      const uses = discordInvite.uses ?? 0;

      // Insert
      const [inserted] = await db
        .insert(trackedInvites)
        .values({
          guildId,
          inviteCode: row.inviteCode,
          channelId: discordInvite.channel?.id ?? "",
          primarySource: row.primarySource,
          secondarySource: row.secondarySource,
          createdBy: row.creatorId,
          uses,
          createdAt: row.createdAt,
        })
        .returning();

      // Add roles
      if (row.roleIds.length > 0 && inserted) {
        for (const roleId of row.roleIds) {
          await db.insert(inviteRoles).values({
            trackedInviteId: inserted.id,
            roleId,
          });
        }
      }

      await updateCachedInvite(guildId, row.inviteCode, uses);
      imported++;
    } catch (err) {
      errors++;
      logger.warn(
        { err, inviteCode: row.inviteCode },
        "Failed to import invite",
      );
    }
  }

  await writeAuditLog(guildId, "bulk_import", interaction.user.id, {
    totalRows: rows.length,
    imported,
    skipped,
    errors,
  });

  const embed = brandEmbed()
    .setTitle("CSV Import Complete")
    .addFields(
      { name: "Total Rows", value: String(rows.length), inline: true },
      { name: "Imported", value: String(imported), inline: true },
      { name: "Skipped", value: String(skipped), inline: true },
      { name: "Errors", value: String(errors), inline: true },
    );

  if (skippedReasons.length > 0) {
    embed.addFields({
      name: "Skip Details",
      value: skippedReasons.slice(0, 10).join("\n"),
    });
  }

  await interaction.editReply({ embeds: [embed] });

  await sendLogMessage(
    interaction.client,
    guildId,
    logEmbed(
      "Bulk Import",
      `<@${interaction.user.id}> imported **${imported}** invites from CSV (${skipped} skipped, ${errors} errors)`,
    ),
  );
}

export default command;

File: src/commands/import.rs
use std::collections::HashMap;

use anyhow::Context as _;
use poise::serenity_prelude as serenity;
use serde_json::json;

use crate::audit::send_log_message;
use crate::csv_transfer;
use crate::embeds;
use crate::invite_tracking::normalize_invite_code;
use crate::models::NewTrackedInvite;
use crate::{Context, Error};

const MAX_CSV_BYTES: u32 = 2 * 1024 * 1024;
const MAX_SKIP_DETAILS: usize = 10;

/// Import existing Discord invites.
#[poise::command(
    slash_command,
    guild_only,
    subcommands("single", "csv"),
    default_member_permissions = "ADMINISTRATOR"
)]
#[allow(clippy::unused_async)]
pub async fn import(_ctx: Context<'_>) -> Result<(), Error> {
    Ok(())
}

/// Import one existing Discord invite.
#[poise::command(slash_command, guild_only)]
#[allow(clippy::too_many_lines)]
pub async fn single(
    ctx: Context<'_>,
    #[description = "Existing invite code or URL"] code: String,
    #[description = "Primary source"]
    #[max_length = 100]
    primary_source: String,
    #[description = "Secondary source"]
    #[max_length = 100]
    secondary_source: String,
    #[description = "Role to assign when someone joins through this invite"] role: Option<
        serenity::Role,
    >,
) -> Result<(), Error> {
    let guild_id = ctx.guild_id().context("command is missing a guild ID")?;
    let guild_id_string = guild_id.to_string();
    let code = normalize_invite_code(&code);
    ctx.data().repository.ensure_guild(&guild_id_string).await?;
    let config = ctx
        .data()
        .repository
        .guild_config(&guild_id_string)
        .await?
        .context("guild configuration disappeared after initialization")?;
    let current_count = ctx
        .data()
        .repository
        .count_invites(&guild_id_string)
        .await?;
    if current_count >= i64::from(config.max_links) {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error(format!(
                    "You have reached the limit of **{}** tracked links.",
                    config.max_links
                )))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }
    if ctx.data().repository.invite_exists(&code).await? {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error(format!(
                    "Invite `{code}` is already being tracked."
                )))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }

    let Some(discord_invite) = guild_id
        .invites(&ctx.serenity_context().http)
        .await?
        .into_iter()
        .find(|invite| invite.code == code)
    else {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error(format!(
                    "Discord does not list an active invite with code `{code}` in this server."
                )))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    };

    let role_ids = role
        .as_ref()
        .map(|role| vec![role.id.to_string()])
        .unwrap_or_default();
    let uses = i32::try_from(discord_invite.uses)
        .context("the invite use count exceeds the supported range")?;
    ctx.data()
        .repository
        .insert_invite(&NewTrackedInvite {
            guild_id: guild_id_string.clone(),
            invite_code: code.clone(),
            channel_id: discord_invite.channel.id.to_string(),
            primary_source: primary_source.clone(),
            secondary_source: secondary_source.clone(),
            created_by: ctx.author().id.to_string(),
            uses,
            created_at: Some(discord_invite.created_at.naive_utc()),
            role_ids,
        })
        .await?;
    ctx.data()
        .cache
        .set_invite(&guild_id_string, &code, discord_invite.uses)
        .await?;
    ctx.data()
        .repository
        .write_audit_log(
            &guild_id_string,
            "invite_imported",
            &ctx.author().id.to_string(),
            Some(json!({
                "inviteCode": code,
                "primarySource": primary_source,
                "secondarySource": secondary_source,
                "uses": uses,
                "roleId": role.as_ref().map(|value| value.id.to_string()),
            })),
        )
        .await?;

    ctx.send(
        poise::CreateReply::default()
            .embed(
                embeds::success()
                    .title("Invite Imported")
                    .description(format!(
                        "Now tracking `https://discord.gg/{code}`.\n\
                         **{primary_source}** → **{secondary_source}**"
                    )),
            )
            .ephemeral(true),
    )
    .await?;
    send_log_message(
        &ctx.serenity_context().http,
        &ctx.data().repository,
        &guild_id_string,
        embeds::log(
            "Invite Imported",
            format!(
                "<@{}> imported `{code}`\n**{primary_source}** → **{secondary_source}**",
                ctx.author().id
            ),
        ),
    )
    .await;
    Ok(())
}

/// Import existing Discord invites from an `InviteAnalytics` CSV export.
#[poise::command(slash_command, guild_only)]
#[allow(clippy::too_many_lines)]
pub async fn csv(
    ctx: Context<'_>,
    #[description = "CSV file to import"]
    #[rename = "file"]
    attachment: serenity::Attachment,
) -> Result<(), Error> {
    if !attachment.filename.to_ascii_lowercase().ends_with(".csv") {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error(
                    "The uploaded file must use the `.csv` extension.",
                ))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }
    if attachment.size > MAX_CSV_BYTES {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error("CSV files must be 2 MiB or smaller."))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }

    let guild_id = ctx.guild_id().context("command is missing a guild ID")?;
    let guild_id_string = guild_id.to_string();
    ctx.defer_ephemeral().await?;

    let bytes = ctx
        .data()
        .attachment_client
        .get(&attachment.url)
        .send()
        .await?
        .error_for_status()?
        .bytes()
        .await?;
    if bytes.len() > usize::try_from(MAX_CSV_BYTES)? {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error("CSV files must be 2 MiB or smaller."))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }

    let rows = csv_transfer::parse_import(&bytes)?;
    if rows.is_empty() {
        ctx.send(
            poise::CreateReply::default()
                .embed(embeds::error(
                    "The CSV file does not contain any invite rows.",
                ))
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }

    ctx.data().repository.ensure_guild(&guild_id_string).await?;
    let config = ctx
        .data()
        .repository
        .guild_config(&guild_id_string)
        .await?
        .context("guild configuration disappeared after initialization")?;
    let mut tracked_count = ctx
        .data()
        .repository
        .count_invites(&guild_id_string)
        .await?;
    let discord_invites = guild_id
        .invites(&ctx.serenity_context().http)
        .await?
        .into_iter()
        .map(|invite| (invite.code.clone(), invite))
        .collect::<HashMap<_, _>>();

    let total = rows.len();
    let mut imported = 0_usize;
    let mut skipped = 0_usize;
    let mut errors = 0_usize;
    let mut skip_details = Vec::new();

    for row in rows {
        if tracked_count >= i64::from(config.max_links) {
            skipped += 1;
            push_skip_detail(
                &mut skip_details,
                format!("`{}`: link limit reached", row.invite_code),
            );
            continue;
        }
        if ctx
            .data()
            .repository
            .invite_exists(&row.invite_code)
            .await?
        {
            skipped += 1;
            push_skip_detail(
                &mut skip_details,
                format!("`{}`: already tracked", row.invite_code),
            );
            continue;
        }
        let Some(discord_invite) = discord_invites.get(&row.invite_code) else {
            skipped += 1;
            push_skip_detail(
                &mut skip_details,
                format!("`{}`: not active in this server", row.invite_code),
            );
            continue;
        };
        let Ok(uses) = i32::try_from(discord_invite.uses) else {
            errors += 1;
            tracing::warn!(
                code = %row.invite_code,
                uses = discord_invite.uses,
                "could not import an invite with an unsupported use count"
            );
            continue;
        };

        let new_invite = NewTrackedInvite {
            guild_id: guild_id_string.clone(),
            invite_code: row.invite_code.clone(),
            channel_id: discord_invite.channel.id.to_string(),
            primary_source: row.primary_source,
            secondary_source: row.secondary_source,
            created_by: row.creator_id,
            uses,
            created_at: row
                .created_at
                .or_else(|| Some(discord_invite.created_at.naive_utc())),
            role_ids: row.role_ids,
        };
        match ctx.data().repository.insert_invite(&new_invite).await {
            Ok(_) => {
                ctx.data()
                    .cache
                    .set_invite(
                        &guild_id_string,
                        &new_invite.invite_code,
                        discord_invite.uses,
                    )
                    .await?;
                imported += 1;
                tracked_count += 1;
            }
            Err(error) => {
                errors += 1;
                tracing::warn!(
                    %error,
                    code = %new_invite.invite_code,
                    "failed to import an invite"
                );
            }
        }
    }

    ctx.data()
        .repository
        .write_audit_log(
            &guild_id_string,
            "bulk_import",
            &ctx.author().id.to_string(),
            Some(json!({
                "total": total,
                "imported": imported,
                "skipped": skipped,
                "errors": errors,
            })),
        )
        .await?;

    let mut embed = embeds::success()
        .title("CSV Import Complete")
        .field("Rows", total.to_string(), true)
        .field("Imported", imported.to_string(), true)
        .field("Skipped", skipped.to_string(), true)
        .field("Errors", errors.to_string(), true);
    if !skip_details.is_empty() {
        embed = embed.field("Skipped Rows", skip_details.join("\n"), false);
    }
    ctx.send(poise::CreateReply::default().embed(embed).ephemeral(true))
        .await?;
    send_log_message(
        &ctx.serenity_context().http,
        &ctx.data().repository,
        &guild_id_string,
        embeds::log(
            "CSV Import Complete",
            format!(
                "<@{}> imported **{imported}** of **{total}** rows. \
                 Skipped: **{skipped}**, errors: **{errors}**.",
                ctx.author().id
            ),
        ),
    )
    .await;
    Ok(())
}

fn push_skip_detail(details: &mut Vec<String>, value: String) {
    if details.len() < MAX_SKIP_DETAILS {
        details.push(value);
    }
}
